File: src/csharp/PiperPlus.Core/Inference/SpeakerEncoder.cs
using System;
using System.IO;
using System.Numerics;
using Microsoft.ML.OnnxRuntime;

namespace PiperPlus.Core.Inference;

/// <summary>
/// Speaker encoder for voice cloning — loads an ECAPA-TDNN ONNX model
/// and extracts speaker embedding vectors from reference audio.
/// </summary>
/// <remarks>
/// <para>
/// Mel spectrogram parameters are unified across all runtimes:
/// sr=16000, n_fft=512, hop=160, n_mels=80, fmin=20, fmax=7600.
/// </para>
/// <para>
/// The extracted embedding (typically 256-d float32) can be passed to
/// <see cref="SynthesisInput.SpeakerEmbedding"/> for voice cloning synthesis.
/// </para>
/// </remarks>
public sealed class SpeakerEncoder : IDisposable
{
    private const int MelSampleRate = 16000;
    private const int MelNFft = 512;
    private const int MelHopLength = 160;
    private const int MelNMels = 80;
    private const float MelFmin = 20f;
    private const float MelFmax = 7600f;

    private readonly InferenceSession _session;
    private bool _disposed;

    /// <summary>
    /// Load a speaker encoder model from an ONNX file.
    /// </summary>
    /// <param name="modelPath">Path to the speaker encoder ONNX model.</param>
    public SpeakerEncoder(string modelPath)
    {
        ArgumentException.ThrowIfNullOrEmpty(modelPath);
        if (!File.Exists(modelPath))
            throw new FileNotFoundException($"Speaker encoder model not found: {modelPath}", modelPath);

        var options = new SessionOptions
        {
            GraphOptimizationLevel = GraphOptimizationLevel.ORT_ENABLE_ALL,
            IntraOpNumThreads = 2,
            InterOpNumThreads = 1,
        };

        _session = new InferenceSession(modelPath, options);
    }

    /// <summary>
    /// Encode audio samples into a speaker embedding vector.
    /// </summary>
    /// <param name="audioSamples">Mono float32 PCM samples.</param>
    /// <param name="sampleRate">Sample rate of the input audio.</param>
    /// <returns>Speaker embedding vector (typically 256-d float32).</returns>
    public float[] Encode(float[] audioSamples, int sampleRate)
    {
        ObjectDisposedException.ThrowIf(_disposed, this);
        ArgumentNullException.ThrowIfNull(audioSamples);

        if (audioSamples.Length == 0)
            throw new ArgumentException("Audio samples cannot be empty.", nameof(audioSamples));

        // Resample to 16kHz if needed
        float[] resampled = sampleRate != MelSampleRate
            ? ResampleLinear(audioSamples, sampleRate, MelSampleRate)
            : audioSamples;

        // Compute mel spectrogram
        float[] mel = ComputeMelSpectrogram(resampled);
        int nFrames = mel.Length / MelNMels;

        if (nFrames == 0)
            throw new ArgumentException("Audio is too short for mel spectrogram computation.", nameof(audioSamples));

        // Create input tensor: [1, n_mels, n_frames]
        using var melTensor = OrtValue.CreateTensorValueFromMemory(
            mel, [1, MelNMels, nFrames]);

        var inputNames = new List<string> { "input" };
        var inputValues = new List<OrtValue> { melTensor };

        string[] outputNames = _session.OutputMetadata.Keys.ToArray();
        if (outputNames.Length == 0) outputNames = ["output"];

        using var runOptions = new RunOptions();
        using var results = _session.Run(runOptions, inputNames, inputValues, outputNames);

        ReadOnlySpan<float> embedding = results[0].GetTensorDataAsSpan<float>();
        return embedding.ToArray();
    }

    /// <summary>
    /// Encode audio from a WAV file into a speaker embedding.
    /// </summary>
    /// <param name="audioPath">Path to a WAV audio file.</param>
    /// <returns>Speaker embedding vector.</returns>
    public float[] EncodeFile(string audioPath)
    {
        ArgumentException.ThrowIfNullOrEmpty(audioPath);
        if (!File.Exists(audioPath))
            throw new FileNotFoundException($"Audio file not found: {audioPath}", audioPath);

        var (samples, sampleRate) = ReadWavFile(audioPath);
        return Encode(samples, sampleRate);
    }

    /// <inheritdoc />
    public void Dispose()
    {
        if (_disposed) return;
        _session.Dispose();
        _disposed = true;
    }

    // ------------------------------------------------------------------
    // Internal: WAV reading
    // ------------------------------------------------------------------

    private static (float[] Samples, int SampleRate) ReadWavFile(string path)
    {
        using var stream = File.OpenRead(path);
        using var reader = new BinaryReader(stream);

        // RIFF header
        string riff = new(reader.ReadChars(4));
        if (riff != "RIFF") throw new InvalidDataException($"Not a WAV file: {path}");
        reader.ReadInt32(); // file size
        string wave = new(reader.ReadChars(4));
        if (wave != "WAVE") throw new InvalidDataException($"Not a WAV file: {path}");

        // Find fmt chunk
        int sampleRate = 0;
        short channels = 0;
        short bitsPerSample = 0;
        short audioFormat = 0;

        while (stream.Position < stream.Length)
        {
            string chunkId = new(reader.ReadChars(4));
            int chunkSize = reader.ReadInt32();

            if (chunkId == "fmt ")
            {
                audioFormat = reader.ReadInt16();
                channels = reader.ReadInt16();
                sampleRate = reader.ReadInt32();
                reader.ReadInt32(); // byte rate
                reader.ReadInt16(); // block align
                bitsPerSample = reader.ReadInt16();
                int remaining = chunkSize - 16;
                if (remaining > 0) reader.ReadBytes(remaining);
            }
            else if (chunkId == "data")
            {
                int numSamples = chunkSize / (bitsPerSample / 8);
                float[] samples;

                if (bitsPerSample == 16)
                {
                    samples = new float[numSamples];
                    for (int i = 0; i < numSamples; i++)
                        samples[i] = reader.ReadInt16() / 32768f;
                }
                else if (bitsPerSample == 32 && audioFormat == 3) // IEEE float
                {
                    samples = new float[numSamples];
                    for (int i = 0; i < numSamples; i++)
                        samples[i] = reader.ReadSingle();
                }
                else
                {
                    throw new NotSupportedException($"Unsupported WAV format: {bitsPerSample}-bit, format={audioFormat}");
                }

                // Convert to mono
                if (channels > 1)
                {
                    int monoLen = samples.Length / channels;
                    float[] mono = new float[monoLen];
                    for (int i = 0; i < monoLen; i++)
                    {
                        float sum = 0;
                        for (int ch = 0; ch < channels; ch++)
                            sum += samples[i * channels + ch];
                        mono[i] = sum / channels;
                    }
                    return (mono, sampleRate);
                }

                return (samples, sampleRate);
            }
            else
            {
                reader.ReadBytes(chunkSize);
            }
        }

        throw new InvalidDataException($"No data chunk found in WAV file: {path}");
    }

    // ------------------------------------------------------------------
    // Internal: Audio processing
    // ------------------------------------------------------------------

    private static float[] ResampleLinear(float[] samples, int fromRate, int toRate)
    {
        double ratio = (double)fromRate / toRate;
        int outputLen = (int)Math.Ceiling(samples.Length / ratio);
        float[] output = new float[outputLen];

        for (int i = 0; i < outputLen; i++)
        {
            double srcPos = i * ratio;
            int idx = (int)srcPos;
            float frac = (float)(srcPos - idx);

            if (idx + 1 < samples.Length)
                output[i] = samples[idx] * (1f - frac) + samples[idx + 1] * frac;
            else if (idx < samples.Length)
                output[i] = samples[idx];
        }

        return output;
    }

    private static float[] ComputeMelSpectrogram(float[] samples)
    {
        float[] melFilters = CreateMelFilterbank();
        float[] window = HannWindow(MelNFft);

        int nFrames = samples.Length >= MelNFft
            ? (samples.Length - MelNFft) / MelHopLength + 1
            : 0;

        int fftBins = MelNFft / 2 + 1;
        float[] melSpec = new float[MelNMels * nFrames];

        for (int frameIdx = 0; frameIdx < nFrames; frameIdx++)
        {
            int start = frameIdx * MelHopLength;

            // Power spectrum via DFT
            float[] powerSpec = new float[fftBins];
            for (int k = 0; k < fftBins; k++)
            {
                float real = 0, imag = 0;
                float freq = -2f * MathF.PI * k / MelNFft;
                for (int n = 0; n < MelNFft; n++)
                {
                    float sample = (start + n < samples.Length)
                        ? samples[start + n] * window[n]
                        : 0f;
                    float angle = freq * n;
                    real += sample * MathF.Cos(angle);
                    imag += sample * MathF.Sin(angle);
                }
                powerSpec[k] = real * real + imag * imag;
            }

            // Apply mel filterbank
            for (int melIdx = 0; melIdx < MelNMels; melIdx++)
            {
                float energy = 0;
                for (int k = 0; k < fftBins; k++)
                    energy += melFilters[melIdx * fftBins + k] * powerSpec[k];

                melSpec[melIdx * nFrames + frameIdx] = MathF.Log(MathF.Max(energy, 1e-10f));
            }
        }

        return melSpec;
    }

    private static float[] HannWindow(int length)
    {
        float[] window = new float[length];
        for (int n = 0; n < length; n++)
            window[n] = 0.5f * (1f - MathF.Cos(2f * MathF.PI * n / length));
        return window;
    }

    private static float[] CreateMelFilterbank()
    {
        int fftBins = MelNFft / 2 + 1;
        float[] filterbank = new float[MelNMels * fftBins];

        float melFmin = HzToMel(MelFmin);
        float melFmax = HzToMel(MelFmax);

        float[] melPoints = new float[MelNMels + 2];
        for (int i = 0; i < melPoints.Length; i++)
            melPoints[i] = melFmin + (melFmax - melFmin) * i / (MelNMels + 1);

        float[] binPoints = new float[melPoints.Length];
        for (int i = 0; i < melPoints.Length; i++)
            binPoints[i] = MelToHz(melPoints[i]) * MelNFft / MelSampleRate;

        for (int m = 0; m < MelNMels; m++)
        {
            // Convert to integer bin indices (matching Python's np.floor().astype(int))
            int left = (int)MathF.Floor(binPoints[m]);
            int center = (int)MathF.Floor(binPoints[m + 1]);
            int right = (int)MathF.Floor(binPoints[m + 2]);

            // Edge case: if the triangle collapses to a single bin, widen it to
            // guarantee a non-zero response (matches Python reference).
            if (left == center && center == right)
            {
                center = Math.Min(center + 1, fftBins - 1);
                right = Math.Min(right + 2, fftBins - 1);
            }
            else if (left == center)
            {
                center = Math.Min(center + 1, fftBins - 1);
            }
            if (center == right)
            {
                right = Math.Min(right + 1, fftBins - 1);
            }

            // Rising slope
            for (int k = left; k < center; k++)
            {
                if (center > left)
                    filterbank[m * fftBins + k] = (float)(k - left) / (center - left);
            }

            // Falling slope
            for (int k = center; k < right; k++)
            {
                if (right > center)
                    filterbank[m * fftBins + k] = (float)(right - k) / (right - center);
            }

            // Ensure center bin always has weight >= 1.0
            if (center < fftBins)
                filterbank[m * fftBins + center] = MathF.Max(filterbank[m * fftBins + center], 1.0f);
        }

        return filterbank;
    }

    private static float HzToMel(float hz) => 2595f * MathF.Log10(1f + hz / 700f);
    private static float MelToHz(float mel) => 700f * (MathF.Pow(10f, mel / 2595f) - 1f);
}

File: src/go/piperplus/speaker_encoder.go
package piperplus

import (
	"encoding/binary"
	"fmt"
	"io"
	"math"
	"os"

	ort "github.com/yalue/onnxruntime_go"
)

// Mel spectrogram parameters — unified across all runtimes.
const (
	melSampleRate = 16000
	melNFFT       = 512
	melHopLength  = 160
	melNMels      = 80
	melFmin       = 20.0
	melFmax       = 7600.0
)

// SpeakerEncoder loads an ECAPA-TDNN speaker encoder ONNX model and
// extracts speaker embeddings from audio. The embedding can be passed
// to SynthesisRequest.SpeakerEmbedding for voice cloning.
type SpeakerEncoder struct {
	session *ort.DynamicAdvancedSession
}

// NewSpeakerEncoder creates a speaker encoder from an ONNX model file.
func NewSpeakerEncoder(modelPath string) (*SpeakerEncoder, error) {
	inputNames := []string{"input"}
	outputNames := []string{"output"}

	sessOpts, err := ort.NewSessionOptions()
	if err != nil {
		return nil, fmt.Errorf("speaker encoder session options: %w", err)
	}
	defer func() { _ = sessOpts.Destroy() }()

	session, err := ort.NewDynamicAdvancedSession(modelPath, inputNames, outputNames, sessOpts)
	if err != nil {
		return nil, fmt.Errorf("speaker encoder load %s: %w", modelPath, err)
	}

	return &SpeakerEncoder{session: session}, nil
}

// Encode extracts a speaker embedding from mono float32 audio samples.
// If sampleRate is not 16000, the audio is resampled via linear interpolation.
func (se *SpeakerEncoder) Encode(audio []float32, sampleRate int) ([]float32, error) {
	if len(audio) == 0 {
		return nil, fmt.Errorf("speaker encoder: empty audio input")
	}

	// Resample to 16kHz if needed
	resampled := audio
	if sampleRate != melSampleRate {
		resampled = resampleLinear(audio, sampleRate, melSampleRate)
	}

	// Compute mel spectrogram
	mel := computeMelSpectrogram(resampled)
	nFrames := len(mel) / melNMels

	if nFrames == 0 {
		return nil, fmt.Errorf("speaker encoder: audio too short for mel spectrogram")
	}

	// Create input tensor: [1, n_mels, n_frames]
	melTensor, err := ort.NewTensor(ort.NewShape(1, int64(melNMels), int64(nFrames)), mel)
	if err != nil {
		return nil, fmt.Errorf("speaker encoder mel tensor: %w", err)
	}
	defer func() { _ = melTensor.Destroy() }()

	// Prepare output
	outputs := []ort.Value{nil}
	inputs := []ort.Value{melTensor}

	if err := se.session.Run(inputs, outputs); err != nil {
		return nil, fmt.Errorf("speaker encoder inference: %w", err)
	}

	// Extract output
	if outputs[0] == nil {
		return nil, fmt.Errorf("speaker encoder: nil output tensor")
	}
	defer func() { _ = outputs[0].Destroy() }()

	outputTensor, ok := outputs[0].(*ort.Tensor[float32])
	if !ok {
		return nil, fmt.Errorf("speaker encoder: unexpected output tensor type")
	}
	rawData := outputTensor.GetData()
	result := make([]float32, len(rawData))
	copy(result, rawData)

	return result, nil
}

// EncodeFile reads a WAV file and extracts a speaker embedding.
func (se *SpeakerEncoder) EncodeFile(path string) ([]float32, error) {
	samples, sampleRate, err := readWavFileForEncoder(path)
	if err != nil {
		return nil, fmt.Errorf("speaker encoder read WAV %s: %w", path, err)
	}
	return se.Encode(samples, sampleRate)
}

// Close destroys the ONNX session.
func (se *SpeakerEncoder) Close() {
	if se.session != nil {
		_ = se.session.Destroy()
		se.session = nil
	}
}

// LoadSpeakerEmbeddingFile reads a pre-computed speaker embedding from a
// raw binary file (little-endian float32 values).
func LoadSpeakerEmbeddingFile(path string) ([]float32, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("read speaker embedding %s: %w", path, err)
	}
	if len(data)%4 != 0 {
		return nil, fmt.Errorf("speaker embedding file size (%d bytes) is not a multiple of 4", len(data))
	}

	floats := make([]float32, len(data)/4)
	for i := range floats {
		bits := binary.LittleEndian.Uint32(data[i*4 : i*4+4])
		floats[i] = math.Float32frombits(bits)
	}
	return floats, nil
}

// readWavFileForEncoder reads a WAV file and returns mono float32 samples.
func readWavFileForEncoder(path string) ([]float32, int, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, 0, err
	}
	defer func() { _ = f.Close() }()

	// Read RIFF header
	var header [12]byte
	if _, err := io.ReadFull(f, header[:]); err != nil {
		return nil, 0, fmt.Errorf("read WAV header: %w", err)
	}
	if string(header[:4]) != "RIFF" || string(header[8:12]) != "WAVE" {
		return nil, 0, fmt.Errorf("not a WAV file: %s", path)
	}

	var sampleRate int
	var channels int
	var bitsPerSample int
	var audioFormat int

	for {
		var chunkHeader [8]byte
		_, err := io.ReadFull(f, chunkHeader[:])
		if err == io.EOF || err == io.ErrUnexpectedEOF {
			break
		}
		if err != nil {
			return nil, 0, fmt.Errorf("read chunk header: %w", err)
		}

		chunkID := string(chunkHeader[:4])
		chunkSize := int(binary.LittleEndian.Uint32(chunkHeader[4:8]))

		switch chunkID {
		case "fmt ":
			if chunkSize < 16 {
				return nil, 0, fmt.Errorf("fmt chunk too small")
			}
			var fmt [16]byte
			if _, err := io.ReadFull(f, fmt[:]); err != nil {
				return nil, 0, err
			}
			audioFormat = int(binary.LittleEndian.Uint16(fmt[0:2]))
			channels = int(binary.LittleEndian.Uint16(fmt[2:4]))
			sampleRate = int(binary.LittleEndian.Uint32(fmt[4:8]))
			bitsPerSample = int(binary.LittleEndian.Uint16(fmt[14:16]))

			remaining := chunkSize - 16
			if remaining > 0 {
				skip := make([]byte, remaining)
				_, _ = io.ReadFull(f, skip)
			}
		case "data":
			numSamples := chunkSize / (bitsPerSample / 8)
			samples := make([]float32, numSamples)

			switch {
			case bitsPerSample == 16:
				for i := 0; i < numSamples; i++ {
					var buf [2]byte
					if _, err := io.ReadFull(f, buf[:]); err != nil {
						break
					}
					val := int16(binary.LittleEndian.Uint16(buf[:]))
					samples[i] = float32(val) / 32768.0
				}
			case bitsPerSample == 32 && audioFormat == 3:
				for i := 0; i < numSamples; i++ {
					var buf [4]byte
					if _, err := io.ReadFull(f, buf[:]); err != nil {
						break
					}
					bits := binary.LittleEndian.Uint32(buf[:])
					samples[i] = math.Float32frombits(bits)
				}
			default:
				return nil, 0, fmt.Errorf("unsupported WAV format: %d-bit, format=%d", bitsPerSample, audioFormat)
			}

			// Convert to mono
			if channels > 1 {
				monoLen := len(samples) / channels
				mono := make([]float32, monoLen)
				for i := 0; i < monoLen; i++ {
					var sum float32
					for ch := 0; ch < channels; ch++ {
						sum += samples[i*channels+ch]
					}
					mono[i] = sum / float32(channels)
				}
				return mono, sampleRate, nil
			}
			return samples, sampleRate, nil
		default:
			// Skip unknown chunk
			skip := make([]byte, chunkSize)
			_, _ = io.ReadFull(f, skip)
		}
	}

	return nil, 0, fmt.Errorf("no data chunk found in WAV file: %s", path)
}

// resampleLinear resamples audio via linear interpolation.
func resampleLinear(samples []float32, fromRate, toRate int) []float32 {
	if fromRate == toRate || len(samples) == 0 {
		return samples
	}

	ratio := float64(fromRate) / float64(toRate)
	outputLen := int(math.Ceil(float64(len(samples)) / ratio))
	output := make([]float32, outputLen)

	for i := 0; i < outputLen; i++ {
		srcPos := float64(i) * ratio
		idx := int(srcPos)
		frac := float32(srcPos - float64(idx))

		if idx+1 < len(samples) {
			output[i] = samples[idx]*(1-frac) + samples[idx+1]*frac
		} else if idx < len(samples) {
			output[i] = samples[idx]
		}
	}

	return output
}

// computeMelSpectrogram computes a log mel spectrogram.
// Returns a flattened [n_mels * n_frames] array (mel-major order).
func computeMelSpectrogram(samples []float32) []float32 {
	melFilters := createMelFilterbank()
	window := hannWindow(melNFFT)

	nFrames := 0
	if len(samples) >= melNFFT {
		nFrames = (len(samples)-melNFFT)/melHopLength + 1
	}

	fftBins := melNFFT/2 + 1
	melSpec := make([]float32, melNMels*nFrames)

	for frameIdx := 0; frameIdx < nFrames; frameIdx++ {
		start := frameIdx * melHopLength

		// Power spectrum via DFT
		powerSpec := make([]float32, fftBins)
		for k := 0; k < fftBins; k++ {
			var real, imag float32
			freq := -2.0 * math.Pi * float64(k) / float64(melNFFT)
			for n := 0; n < melNFFT; n++ {
				var sample float32
				if start+n < len(samples) {
					sample = samples[start+n] * window[n]
				}
				angle := freq * float64(n)
				real += sample * float32(math.Cos(angle))
				imag += sample * float32(math.Sin(angle))
			}
			powerSpec[k] = real*real + imag*imag
		}

		// Apply mel filterbank
		for melIdx := 0; melIdx < melNMels; melIdx++ {
			var energy float32
			for k := 0; k < fftBins; k++ {
				energy += melFilters[melIdx*fftBins+k] * powerSpec[k]
			}
			if energy < 1e-10 {
				energy = 1e-10
			}
			melSpec[melIdx*nFrames+frameIdx] = float32(math.Log(float64(energy)))
		}
	}

	return melSpec
}

func hannWindow(length int) []float32 {
	window := make([]float32, length)
	for n := 0; n < length; n++ {
		window[n] = 0.5 * (1 - float32(math.Cos(2*math.Pi*float64(n)/float64(length))))
	}
	return window
}

func createMelFilterbank() []float32 {
	fftBins := melNFFT/2 + 1
	filterbank := make([]float32, melNMels*fftBins)

	melFminVal := hzToMel(melFmin)
	melFmaxVal := hzToMel(melFmax)

	melPoints := make([]float32, melNMels+2)
	for i := range melPoints {
		melPoints[i] = melFminVal + (melFmaxVal-melFminVal)*float32(i)/float32(melNMels+1)
	}

	binPoints := make([]float32, len(melPoints))
	for i, m := range melPoints {
		binPoints[i] = melToHz(m) * float32(melNFFT) / float32(melSampleRate)
	}

	for m := 0; m < melNMels; m++ {
		// Convert to integer bin indices (matching Python's np.floor().astype(int))
		left := int(math.Floor(float64(binPoints[m])))
		center := int(math.Floor(float64(binPoints[m+1])))
		right := int(math.Floor(float64(binPoints[m+2])))

		// Edge case: if the triangle collapses to a single bin, widen it to
		// guarantee a non-zero response (matches Python reference).
		if left == center && center == right {
			center = min(center+1, fftBins-1)
			right = min(right+2, fftBins-1)
		} else if left == center {
			center = min(center+1, fftBins-1)
		}
		if center == right {
			right = min(right+1, fftBins-1)
		}

		// Rising slope
		for k := left; k < center; k++ {
			if center > left {
				filterbank[m*fftBins+k] = float32(k-left) / float32(center-left)
			}
		}

		// Falling slope
		for k := center; k < right; k++ {
			if right > center {
				filterbank[m*fftBins+k] = float32(right-k) / float32(right-center)
			}
		}

		// Ensure center bin always has weight >= 1.0
		if center < fftBins {
			if filterbank[m*fftBins+center] < 1.0 {
				filterbank[m*fftBins+center] = 1.0
			}
		}
	}

	return filterbank
}

func hzToMel(hz float32) float32 {
	return 2595 * float32(math.Log10(float64(1+hz/700)))
}

func melToHz(mel float32) float32 {
	return 700 * (float32(math.Pow(10, float64(mel)/2595)) - 1)
}
